orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,503 @@
+		nextSeekInfo(startSeq),
+		int32(0),
+		uint64(0),
+		util.ComputeSHA256(p.Config.TLSCert),
+	)
+}
+
+func (p *AttestationPuller) closeEndpoints() {
+	// Close the streams and grpc connections
+	for _, oc := range p.openConnections {
+		oc.stream.cancelFunc()
+		if oc.conn != nil {
+			oc.conn.Close()
+		}
+	}
+	p.openConnections = nil
+}
+
+func (p *AttestationPuller) Close() {
+	p.closeEndpoints()
+}
+
+func (p *AttestationPuller) pullAttestationBlock(ec EndpointCriteria, seq uint64, env *common.Envelope) (*orderer.BlockAttestation, error) {
+
+	// The connection should be via closed via Close() method of AttestationPuller
+	conn, err := p.Config.Dialer.Dial(ec)
+	if err != nil {
+		return nil, err
+	}
+
+	stream, err := newImpatientStream(conn, p.Config.FetchTimeout, env)
+	if err != nil {
+		return nil, err
+	}
+
+	// append the stream to bf.openConnections, so that Close can close all streams
+	// dial connections
+	cni := attestationConnectionInfo{stream: stream, conn: conn}
+	p.openConnections = append(p.openConnections, cni)
+
+	resp, err := stream.Recv()
+	status := resp.GetStatus()
+	if status != common.Status_SUCCESS {
+		p.Logger.Errorf("error in rpc response %v", status)
+		if status != common.Status_UNKNOWN {
+			return nil, errors.Errorf("error in rpc response: %v", status)
+		}
+	}
+	attestation := resp.GetBlockAttestation()
+	return attestation, err
+}
+
+type impatientStreamCreator func(in *common.Envelope) (*impatientAttestationStream, error)
+
+// impatientAttestationStream aborts the stream if it waits for too long for a message.
+type impatientAttestationStream struct {
+	waitTimeout time.Duration
+	orderer.BlockAttestations_BlockAttestationsClient
+	cancelFunc func()
+}
+
+func (stream *impatientAttestationStream) abort() {
+	stream.cancelFunc()
+}
+
+// Recv blocks until a response is received from the stream or the
+// timeout expires.
+func (stream *impatientAttestationStream) Recv() (*ab.BlockAttestationResponse, error) {
+
+	// Initialize a timeout to cancel the stream when it expires
+	timeout := time.NewTimer(stream.waitTimeout)
+	defer timeout.Stop()
+
+	responseChan := make(chan attestationErrorAndResponse, 1)
+
+	// receive waitGroup ensures the goroutine below exits before
+	// this function exits.
+	var receive sync.WaitGroup
+	receive.Add(1)
+	defer receive.Wait()
+
+	go func() {
+		defer receive.Done()
+		resp, err := stream.BlockAttestations_BlockAttestationsClient.Recv()
+		responseChan <- attestationErrorAndResponse{err: err, resp: resp}
+	}()
+
+	select {
+	case <-timeout.C:
+		stream.cancelFunc()
+		return nil, errors.Errorf("didn't receive a response within %v", stream.waitTimeout)
+	case respAndErr := <-responseChan:
+		return respAndErr.resp, respAndErr.err
+	}
+}
+
+// newImpatientStream returns a ImpatientStreamCreator that creates impatientStreams.
+func newImpatientStream(conn *grpc.ClientConn, waitTimeout time.Duration, in *common.Envelope) (*impatientAttestationStream, error) {
```

Comment:
`NewImpatientStream` exists in the same package so let's call this `newImpatientAttestationStream`?

Reply:
this is a slightly different function since it connects to a different grpc service